Bugfix/cam rev9

MarcelCaron-NOAA opened the pull request:
## Pull Request Testing ##

- [x] Describe testing already performed for this Pull Request:</br>

This PR addresses issue 86,.1-10, 34-36 noted in the Google Doc titled "EVS v1.0 issues".  Information about its cause and fix:

* 86 - When envir is set to a bogus directory name (to test how jobs handle missing com data), some cam prep and stats jobs either (1) fail with a FATAL ERROR, (2) complete but without WARNINGs, or (3) both.  Various WARNING messages are missing, or misspelled, or located inside of mail messages.  Also some jobs run METplus when needed files are missing.  Finally some jobs output FATAL ERROR when no job cards are created.  All of these contribute to the issues handling the bogus directory name test. _Fix:_  Correct or add WARNING messages.  Fix error handling, which includes replacing some err_exit cases with booleans that tell jobs whether or not to skip METplus.  Change some FATAL ERROR cases to NOTE where appropriate.

A representative subset of the affected jobs was tested.

---
- [x] Recommend testing for the reviewer(s) to perform, including the location of input datasets, and any additional instructions:</br>

For the nam_firewxnest job:

> Do all tests using SENDMAIL=NO:
> 
> Test normally (using "prod" input). Check to see that the -e statements are only on filetimes that exist. Run at least twice:
> qsub -v vhr=00
> qsub -v vhr=01
> 
> Then, in the dev driver, set COMINnam to
> 
> export COMINnam=/lfs/h1/ops/pppp/com/nam/${nam_ver}
> 
> This will fake the NAM directory. Then run the dev driver again.
> 
> Only one run here will suffice: qsub -v vhr=00. You may do two for a more rigorous test at another vhr.
> (3 runs total)

For other jobs:
### (1) Set up jobs 
•	symlink the EVS_fix directory locally as "fix"
•	In all driver scripts, point all exports of "HOMEevs" to your EVS test directory
•	In all driver scripts, change COMIN to the parallel directory (`/lfs/h2/emc/vpppg/noscrub/emc.vpppg/$NET/$evs_ver_2d`)
•	In all driver scripts, set SENDMAIL to "NO"
•	In all driver scripts, set envir to "pppp"

### (2) Test Jobs
Use the following to submit ...

- ... jevs_cam_namnest_precip_stats: `qsub –v vhr=19,VDATE=$(date -d "-3 days" +"%Y%m%d")` 
- ... jevs_cam_nam_firewxnest_grid2obs_stats: see quoted instructions above
- ... all other jobs: `qsub –v vhr=00 `

From your EVS testing directory, the following scripts should be run using the setup in step (1):
```
dev/drivers/scripts/prep/cam/jevs_cam_namnest_precip_prep.sh
dev/drivers/scripts/prep/cam/jevs_cam_namnest_severe_prep.sh
```
```
dev/drivers/scripts/plots/cam/jevs_cam_namnest_grid2obs_stats.sh
dev/drivers/scripts/plots/cam/jevs_cam_namnest_precip_stats.sh
dev/drivers/scripts/plots/cam/jevs_cam_namnest_radar_stats.sh
dev/drivers/scripts/plots/cam/jevs_cam_namnest_severe_stats.sh
dev/drivers/scripts/plots/cam/jevs_cam_namnest_snowfall_stats.sh
dev/drivers/scripts/plots/cam/jevs_cam_href_grid2obs_stats.sh
dev/drivers/scripts/plots/cam/jevs_cam_href_precip_stats.sh
dev/drivers/scripts/plots/cam/jevs_cam_href_spcoutlook_stats.sh
dev/drivers/scripts/stats/cam/jevs_cam_nam_firewxnest_grid2obs_stats.sh
```

_[Total: 2 prep jobs; 9 stats jobs]_


### (3) Review

- check for FATAL ERROR: `grep -l "FATAL ERROR" $outfile`
- check for no WARNINGs: `grep -L "WARNING" $outfile`

No jobs should output any FATAL ERRORs.  All jobs should output WARNINGs indicating missing files.

---
- [x] Has the code been checked to ensure that no errors occur during the execution? **YES**

---
- [x] Do these updates/additions include sufficient testing updates? **YES**

---
- [x] Please complete this pull request review by **02/14/2024**.</br>

---
## Pull Request Checklist ##

- [ ] Review the source issue metadata (required labels, projects, and milestone).
- [ ] Complete the PR description above.
- [ ] Ensure the PR title matches the feature branch name.
- [ ] Check the following:
- [ ]  Instructions provided on how to run
- [ ]  Developer's name is replaced by ${user} where necessary throughout the code
- [ ]  Check that the ecf file has all the proper definitions of variables
- [ ]  Check that the jobs file has all the proper settings of COMIN and COMOUT and other input variables
- [ ]  Check to see that the output directory structure is followed
- [ ]  Be sure that you are not using MET utilities outside the METplus wrapper structure

- [ ] After submitting the PR, select **Development** issue with the original issue number.
- [ ] After the PR is approved, merge your changes. If permissions do not allow this, request that the reviewer do the merge.
- [ ] Close the linked issue.

Commit:
Fix issue 86
Change FATAL ERROR to NOTE when there are no job cards.

Commit:
Merge branch 'develop' into bugfix/cam_rev9

Commit:
Fix item 86.1-10, 34-36
Source variables from href file checks + use verify booleans to
conditionally run METplus

Commit:
Fix item 86.1-10, 34-36
Correct WARNING spelling + Add WARNING label + Add WARNING message
outside of mail message block

Commit:
Fix 86.1-10, 34-36
Change err_exit to WARNING when missing data + set verify booleans to
pass information about whether or not to run METplus back to parent

Commit:
Fix item 86.1-10, 34-36
Correct/Add WARNING label

Commit:
Merge branch 'develop' into bugfix/cam_rev9

PerryShafran-NOAA commented:
@MarcelCaron-NOAA Please incorporate the firexnest changes from #444 into this PR.  My PR for 444 was specifically for firewxnest, and also there was a change to the if blocks, as I reordered the blocks so that it's only looking for files for model cycles that exist.  I also added a WARNING statement to list the forecast hours that are missing.  Let me know if you have any questions about it.  

I would like the firewxnest forecast tested for both vhr=00 and vhr=01 to see if this is working properly.  

@ShelleyMelchior-NOAA for awareness

Commit:
Fix item 86.9
No longer search for files at times that don't exist
Collect missing fhrs + print single WARNING if any are missing

MarcelCaron-NOAA commented:
I've backed out my edits to nam_firewxnest and added those made in [PR#444](https://github.com/NOAA-EMC/EVS/pull/444).  If there are no issues with those additions to `scripts/stats/cam/exevs_cam_nam_firewxnest_grid2obs_stats.sh`, then we should be all set!

CC @ShelleyMelchior-NOAA @PerryShafran-NOAA

MarcelCaron-NOAA commented:
I've also added the separate instructions for testing jevs_cam_nam_firewxnest in the PR description.

ShelleyMelchior-NOAA commented:
Fantastic. Thanks for the coordniation, @MarcelCaron-NOAA and @PerryShafran-NOAA . I will get started on this PR this morning.

PerryShafran-NOAA commented:
@MarcelCaron-NOAA @ShelleyMelchior-NOAA Looking at your changes from earlier, I feel that you might want to add those back in.  They were the WARNING messages when SENDMAIL=NO.  Do you think they should still be in there?

PerryShafran-NOAA commented:
@ShelleyMelchior-NOAA @MarcelCaron-NOAA I'd also like firewxnest to be tested with envir=prod as normal.  I do think the envir=pppp test is critical as well, but since I made edits, I need to test with normal data as well.

MarcelCaron-NOAA commented:
@PerryShafran-NOAA I backed out the change I made (printing a warning message if SENDMAIL=NO) because your changes will do the same thing (increment "fcst_miss" whether or not SENDMAIL=NO, then print a WARNING later if fcst_miss>0).

PerryShafran-NOAA commented:
@MarcelCaron-NOAA Oh!  that makes sense.  But your original message printed a warning for individual files, whereas my warning is a single statement that lists all forecasts that are missing.  

@ShelleyMelchior-NOAA Does this make sense to you also?

AliciaBentley-NOAA commented:
@MarcelCaron-NOAA @PerryShafran-NOAA @ShelleyMelchior-NOAA As long as the individual files that are missing is written out in some way, that is fine. I disagree with some of the advice that has been given lately about "one email" being sent, etc etc. Emails won't be sent it ops. As long as the individual files that are missing make it to a log file in some capacity, we are OK. Thanks!

PerryShafran-NOAA commented:
@MarcelCaron-NOAA Based on Alicia's comment, you should put that WARNING statement back into the code.  It needs to be there. 

@ShelleyMelchior-NOAA @AliciaBentley-NOAA

PerryShafran-NOAA commented:
@MarcelCaron-NOAA I will be dividing the work with Shelley on this PR.  She will be doing all the data denial experiments (changing envir to pppp), while I will be doing nominal experiments (with regular data available), to test both ways.  I have started on the prep step.  You can find the .o files here:  /lfs/h2/emc/vpppg/noscrub/perry.shafran/pr445test/EVS/dev/drivers/scripts/prep/cam

I did not set envir to pppp, so these are runs with all data available.  

@ShelleyMelchior-NOAA

PerryShafran-NOAA commented:
@MarcelCaron-NOAA The nominal runs for the prep step has completed.

The output can be found here:  /lfs/h2/emc/vpppg/noscrub/perry.shafran/evs/v1.0/prep/cam/____.20240211

Commit:
Fix item 86.9
Add WARNING statement that triggers if SENDMAIL=NO

ShelleyMelchior-NOAA commented:
The data denial prep jobs have completed.
logfiles:
`/lfs/h2/emc/ptmp/shelley.melchior/EVS_out/jevs_cam_namnest_precip_prep.o132628580`
`/lfs/h2/emc/ptmp/shelley.melchior/EVS_out/jevs_cam_namnest_severe_prep.o132628594`

PerryShafran-NOAA commented:
@MarcelCaron-NOAA The nominal stats jobs are underway and their results can be found here:  /lfs/h2/emc/vpppg/noscrub/perry.shafran/pr445test/EVS/dev/drivers/scripts/stats/cam

I did not do the firewxnest job yet as we await the addition of that line outside the SENDMAIL block that  you had in before.

MarcelCaron-NOAA commented:
> The data denial prep jobs have completed. logfiles: `/lfs/h2/emc/ptmp/shelley.melchior/EVS_out/jevs_cam_namnest_precip_prep.o132628580` `/lfs/h2/emc/ptmp/shelley.melchior/EVS_out/jevs_cam_namnest_severe_prep.o132628594`

Data denial test (prep):
Both prep jobs outputted WARNINGs for missing files.  Neither prep job exited with FATAL ERROR.
✔️ Prep jobs pass data denial tests

MarcelCaron-NOAA commented:
> @MarcelCaron-NOAA Based on Alicia's comment, you should put that WARNING statement back into the code. It needs to be there.
> 
> @ShelleyMelchior-NOAA @AliciaBentley-NOAA

@PerryShafran-NOAA The else conditional block was reinserted so that individual missing firewx files are printed when SENDMAIL=NO 👍

PerryShafran-NOAA commented on scripts/stats/cam/exevs_cam_nam_firewxnest_grid2obs_stats.sh:
```diff
@@ -39,26 +40,38 @@ do
         cat mailmsg | mail -s "$subject" $MAILTO
+       else
+        echo "WARNING: The NAM Firewx file is missing for valid date ${VDATE}. METplus will not run."
        fi
```
Hi, @MarcelCaron-NOAA ,

Thank you for putting this line back into the code.  However, Alicia has stated that the full path needs to be in this WARNING statement.  Thus, it should read something like this:

`"WARNING: File $COMINnam/nam.${aday}/nam.t${acyc}z.${regionnest}.${outtyp}${fhr}.tm00.grib2 is missing"  `

Also, METplus will run if a single file is missing, but it won't run if many files are missing.  Thus, your warning message should not have "METplus will not run"  if a single file is missing.  It shouldn't run if all files are missing, though.  That is taken care of in a message below.  

@ShelleyMelchior-NOAA

MarcelCaron-NOAA replied:
Okay let me add in the full path.  The mail message also says "METplus will not run."  Should I remove that line too?

AliciaBentley-NOAA replied:
@MarcelCaron-NOAA I think you can leave that. My bigger concern is this is an "else" statement that depends on SENDMAIL. This statement should be written no matter what SENDMAIL is set to. Can we separate these WARNINGs from SENDMAIL entirely? @PerryShafran-NOAA @ShelleyMelchior-NOAA

PerryShafran-NOAA replied:
Yes please, remove the "METplus will not run" part of that line at the least.

MarcelCaron-NOAA replied:
@PerryShafran-NOAA Full path added 👍

PerryShafran-NOAA replied:
@MarcelCaron-NOAA Yes, in that case you should just place that line after the let fcstmiss line and it would not be part of the SENDMAIL block.

PerryShafran-NOAA replied:
@MarcelCaron-NOAA Thanks for adding it the full path.  Now per Alicia's instructions, you need to move it out of the SENDMAIL block entirely.  I would place it after line 50, after the let fcstmiss statement.

PerryShafran-NOAA replied:
Excellent, thanks!

Commit:
Fix item 86.9
Add full path for missing file logs

MarcelCaron-NOAA commented:
> /lfs/h2/emc/vpppg/noscrub/perry.shafran/pr445test/EVS/dev/drivers/scripts/prep/cam

@PerryShafran-NOAA,

Nominal test (prep):
✔️ jevs_cam_namnest_precip_prep ran successfully and output files look good.
❔ jevs_cam_namnest_severe_prep log file is missing?

PerryShafran-NOAA commented:
@MarcelCaron-NOAA Whoops!  I tried to delete a file with an error but I screwed up.  Let me try this again.

PerryShafran-NOAA commented:
@MarcelCaron-NOAA Check the .o file again.

Commit:
Fix item 86.9
Move WARNING message to outside of SENDMAIL if block

Commit:
Fix item 86.9
Move WARNING message to before the SENDMAIL if block

Commit:
Fix item 86.9
Remove "METplus will not run" part of mail message about missing firewx
files.

PerryShafran-NOAA commented:
OK I kicked off the nam_firewx run with all the other stats jobs already running, for the nominal condition.  All the .o files are here: 

 /lfs/h2/emc/vpppg/noscrub/perry.shafran/pr445test/EVS/dev/drivers/scripts/stats/cam

Also don't forget to look at the prep job where the .o file was missing earlier; it's there now.

MarcelCaron-NOAA commented:
> > /lfs/h2/emc/vpppg/noscrub/perry.shafran/pr445test/EVS/dev/drivers/scripts/prep/cam
> 
> @PerryShafran-NOAA,
> 
> Nominal test (prep): ✔️ jevs_cam_namnest_precip_prep ran successfully and output files look good. ❔ jevs_cam_namnest_severe_prep log file is missing?

Thanks @PerryShafran-NOAA!
✔️ jevs_cam_namnest_severe_prep ran successfully and output files look good

ShelleyMelchior-NOAA commented:
All of the data denial stats jobs have been run per the supplied instructions. No hits when I grep on FATAL. Lots of WARNINGs, as expected.
```
-rw-r--r-- 1 shelley.melchior emc   1150092 Feb 12 17:17 jevs_cam_namnest_radar_stats.o132632334
-rw-r--r-- 1 shelley.melchior emc   1009181 Feb 12 17:17 jevs_cam_namnest_severe_stats_00.o132632337
-rw-r--r-- 1 shelley.melchior emc   1749105 Feb 12 17:18 jevs_cam_namnest_snowfall_stats.o132632338
-rw-r--r-- 1 shelley.melchior emc   5263780 Feb 12 17:18 jevs_cam_namnest_precip_stats.o132632325
-rw-r--r-- 1 shelley.melchior emc   5277789 Feb 12 17:19 jevs_cam_namnest_grid2obs_stats.o132632324
-rw-r--r-- 1 shelley.melchior emc    937134 Feb 12 17:24 jevs_cam_href_grid2obs_stats.o132632745
-rw-r--r-- 1 shelley.melchior emc    945258 Feb 12 17:24 jevs_cam_href_precip_stats.o132632746
-rw-r--r-- 1 shelley.melchior emc    937263 Feb 12 17:24 jevs_cam_href_spcoutlook_stats.o132632751
-rw-r--r-- 1 shelley.melchior emc    947177 Feb 12 17:29 jevs_cam_nam_firewxnest_grid2obs_stats_00.o132633152
[dlogin01 /lfs/h2/emc/ptmp/shelley.melchior/EVS_out]$ grep -i FATAL jevs_cam_nam*stats*
[dlogin01 /lfs/h2/emc/ptmp/shelley.melchior/EVS_out]$ 

```

MarcelCaron-NOAA commented:
@ShelleyMelchior-NOAA Agreed.

✔️ dev/drivers/scripts/plots/cam/jevs_cam_namnest_grid2obs_stats.sh
✔️ dev/drivers/scripts/plots/cam/jevs_cam_namnest_precip_stats.sh
✔️ dev/drivers/scripts/plots/cam/jevs_cam_namnest_radar_stats.sh
✔️ dev/drivers/scripts/plots/cam/jevs_cam_namnest_severe_stats.sh
✔️ dev/drivers/scripts/plots/cam/jevs_cam_namnest_snowfall_stats.sh
✔️ dev/drivers/scripts/plots/cam/jevs_cam_href_grid2obs_stats.sh
✔️ dev/drivers/scripts/plots/cam/jevs_cam_href_precip_stats.sh
✔️ dev/drivers/scripts/plots/cam/jevs_cam_href_spcoutlook_stats.sh
👍 dev/drivers/scripts/plots/cam/jevs_cam_nam_firewxnest_stats.sh

Data denial tests (stats):
All stats jobs outputted WARNINGs for missing files. None of the stats jobs exited with FATAL ERROR.

No further concerns about the data denial tests from me, but I'll let @PerryShafran-NOAA weigh in on the firewxnest output.

PerryShafran-NOAA commented:
@ShelleyMelchior-NOAA Did you run the firewxnest before or after Marcel's latest changes?  I don't see the WARNING statement for each missing file.

ShelleyMelchior-NOAA commented:
Let me re-run ... it was after Marcel's change, but I forgot to run `git pull`. Stand by.

PerryShafran-NOAA commented:
@MarcelCaron-NOAA I finished the vhr=00 and vhr=01 runs of nam_firewxnest for the nominal jobs.  They look good.  I'm going to try to see what happens if there is *one* missing file somewhere.

ShelleyMelchior-NOAA commented:
@PerryShafran-NOAA , @MarcelCaron-NOAA , the firewx stats data denial re-run has completed.
`/lfs/h2/emc/ptmp/shelley.melchior/EVS_out/jevs_cam_nam_firewxnest_grid2obs_stats_00.o132634416`

PerryShafran-NOAA commented:
OK, for my latest run, I set VDATE=20240206, and that means some files are available, older files are not.  I did see a successful warning for the single file that was missing, but METplus still runs with the available files valid at that date.  

@ShelleyMelchior-NOAA @MarcelCaron-NOAA

PerryShafran-NOAA commented:
@ShelleyMelchior-NOAA Your latest data denial run looks good!  I see the individual file warnings.  

@MarcelCaron-NOAA

PerryShafran-NOAA commented:
@MarcelCaron-NOAA All jobs in the nominal case for the non-firewx runs are done except for the href precip.  The .o files are here:  /lfs/h2/emc/vpppg/noscrub/perry.shafran/pr445test/EVS/dev/drivers/scripts/stats/cam

You can find all the output here:  /lfs/h2/emc/vpppg/noscrub/perry.shafran/evs/v1.0/stats/cam/atmos.20240211

@ShelleyMelchior-NOAA

MarcelCaron-NOAA commented:
@ShelleyMelchior-NOAA @PerryShafran-NOAA The firewx nest data denial re-run looks good to me too!

MarcelCaron-NOAA commented:
@PerryShafran-NOAA All nominal runs but two href jobs have completed. 

All of the following jobs completed cleanly.  Any WARNINGs are expected.  All of their outputs appear normal.
✔️ dev/drivers/scripts/plots/cam/jevs_cam_namnest_grid2obs_stats.sh
✔️ dev/drivers/scripts/plots/cam/jevs_cam_namnest_precip_stats.sh
✔️ dev/drivers/scripts/plots/cam/jevs_cam_namnest_radar_stats.sh
✔️ dev/drivers/scripts/plots/cam/jevs_cam_namnest_severe_stats.sh
✔️ dev/drivers/scripts/plots/cam/jevs_cam_namnest_snowfall_stats.sh
✔️ dev/drivers/scripts/plots/cam/jevs_cam_href_spcoutlook_stats.sh
✔️ dev/drivers/scripts/plots/cam/jevs_cam_nam_firewxnest_stats.sh - v00 and v01 output are comparable to the parallel output

The following are still running and will be reviewed later:

- dev/drivers/scripts/plots/cam/jevs_cam_href_grid2obs_stats.sh
- dev/drivers/scripts/plots/cam/jevs_cam_href_precip_stats.sh

PerryShafran-NOAA commented:
@MarcelCaron-NOAA Had to be sure I submitted the href grid2obs stats job.  Turned out I ran the hrrr grid2obs stats job instead!  

Please check out the href precip job, that one looks to be finished.

MarcelCaron-NOAA commented:
> @MarcelCaron-NOAA Had to be sure I submitted the href grid2obs stats job. Turned out I ran the hrrr grid2obs stats job instead!
> 
> Please check out the href precip job, that one looks to be finished.

Thanks @PerryShafran-NOAA!

✔️ dev/drivers/scripts/plots/cam/jevs_cam_href_precip_stats.sh completed cleanly, and its output appears normal.

Confirmed that jevs_cam_href_grid2obs_stats is now running.  I'll review it when it completes (~3 hours)

PerryShafran-NOAA commented:
@MarcelCaron-NOAA Please sync your fork when you get a chance.

Commit:
Merge branch 'develop' into bugfix/cam_rev9

PerryShafran-NOAA commented:
@MarcelCaron-NOAA The job is still running, but please check the href grid2obs stats output when you have a chance and see if it looks good thus far.

MarcelCaron-NOAA commented:
@PerryShafran-NOAA Good news, the job is running cleanly so far.  I expect it will be finished around 5 - 5:15 EST

PerryShafran-NOAA commented:
@MarcelCaron-NOAA The job is complete.  Can you take a look at the .o file and output please?

MarcelCaron-NOAA commented:
@PerryShafran-NOAA:

✔️ dev/drivers/scripts/plots/cam/jevs_cam_href_grid2obs_stats.sh completed cleanly, and its output appears normal.

All data denial and nominal tests completed successfully.  No further concerns about this PR from me.

CC: @AliciaBentley-NOAA @ShelleyMelchior-NOAA

PerryShafran-NOAA reviewed (approved):
Changes work as expected.  Approved for merge.
